state/runtime/tracer/calltracer/call_tracer.go

```diff
@@ -0,0 +1,156 @@
+package calltracer
+
+import (
+	"math/big"
+	"sync"
+
+	"github.com/0xPolygon/polygon-edge/helper/hex"
+	"github.com/0xPolygon/polygon-edge/state/runtime/tracer"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+var (
+	callTypes = map[int]string{
+		0: "CALL",
+		1: "CALLCODE",
+		2: "DELEGATECALL",
+		3: "STATICCALL",
+		4: "CREATE",
+		5: "CREATE2",
+	}
+)
+
+type Call struct {
+	Type    string  `json:"type"`
+	From    string  `json:"from"`
+	To      string  `json:"to"`
+	Value   string  `json:"value,omitempty"`
+	Gas     string  `json:"gas"`
+	GasUsed string  `json:"gasUsed"`
+	Input   string  `json:"input"`
+	Output  string  `json:"output"`
+	Calls   []*Call `json:"calls,omitempty"`
+
+	parent   *Call
+	startGas uint64
+}
+
+type CallTracer struct {
+	call               *Call
+	activeCall         *Call
+	activeGas          uint64
+	activeAvailableGas uint64
+
+	cancelLock *sync.Mutex
```

Comment:
No need to instantiate mutex, not sure are there any benefits if it is a pointer? Also probably it is better to use an `RWMutex`.
```suggestion
	cancelLock sync.RWMutex
```

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/2001/commits/786e10e03573e7a20dd6ec34eb45ff613da5909c